File: cpp/Graph/L_noOfIncPathOnGrid.cpp
#include <bits/stdc++.h>
using namespace std;

int countPaths(vector<vector<int>> &grid)
{
    int rows = grid.size();
    int cols = grid[0].size();
    int mod = 1e9 + 7;

    vector<vector<int>> dp(rows, vector<int>(cols, -1));

    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < cols; j++)
        {
            if (dp[i][j] == -1)
            {
                helper(i, j, grid, dp, -1);
            }
        }
    }

    long long pathSum = 0;
    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < cols; j++)
        {
            pathSum = (pathSum + dp[i][j]) % mod;
        }
    }
    return static_cast<int>(pathSum);
}

int helper(int row, int col, vector<vector<int>> &grid, vector<vector<int>> &dp, int parent)
{
    int rows = grid.size();
    int cols = grid[0].size();
    int mod = 1e9 + 7;

    if (row < 0 || row >= rows || col < 0 || col >= cols || grid[row][col] <= parent)
    {
        return 0;
    }

    if (dp[row][col] != -1)
    {
        return dp[row][col];
    }

    int down = helper(row + 1, col, grid, dp, grid[row][col]) % mod;
    int up = helper(row - 1, col, grid, dp, grid[row][col]) % mod;
    int right = helper(row, col + 1, grid, dp, grid[row][col]) % mod;
    int left = helper(row, col - 1, grid, dp, grid[row][col]) % mod;

    return dp[row][col] = (down + up + right + left + 1) % mod;
}

File: Python/Graphs/L_noOfIncPathOnGrid.py
from functools import lru_cache


def countPaths(grid) -> int:
        rows = len(grid)
        cols = len(grid[0])
        mod = 10 ** 9 + 7

        @lru_cache(None)
        def count(row, col):
            res = 1
            for dx, dy in [[row, col + 1], [row, col - 1], [row + 1, col], [row - 1, col]]:
                if 0 <= dx < rows and 0 <= dy < cols and grid[dx][dy] > grid[row][col]:
                    res += count(dx, dy) % mod
            return res

        count_values = []
        for i in range(rows):
            for j in range(cols):
                count_values.append(count(i, j))
        path_sum = sum(count_values) % mod

        return path_sum